### PR TITLE
feat: implement TryFrom trait to `SignedAmount` and `Amount`

### PR DESCRIPTION
**Pull request by 0xkrieger**
Closes: #2245 

Adds `TryFrom<SignedAmount> for Amount` and `TryFrom<Amount> for Amount` in units module

### CODE REVIEW

File: units/src/amount.rs [hunk 1]
```diff
@@ -957,6 +957,15 @@ impl FromStr for Amount {
     fn from_str(s: &str) -> Result<Self, Self::Err> { Amount::from_str_with_denomination(s) }
 }
 
+impl TryFrom<SignedAmount> for Amount {
+    type Error = ParseAmountError;
```
Comment (Kixunil) [inline units/src/amount.rs:961]: This needs to be `OutOfRangeError`

File: units/src/amount.rs [hunk 2]
```diff
@@ -957,6 +957,15 @@ impl FromStr for Amount {
...
+
+    fn try_from(value: SignedAmount) -> Result<Self, Self::Error> {
+        let unsigned_amount = value.to_unsigned()?;
+        Ok(unsigned_amount)
```
Comment (Kixunil) [inline units/src/amount.rs:965]: Conversion shouldn't be required here.

File: units/src/amount.rs [hunk 3]
```diff
@@ -1334,6 +1343,14 @@ impl FromStr for SignedAmount {
     fn from_str(s: &str) -> Result<Self, Self::Err> { SignedAmount::from_str_with_denomination(s) }
 }
 
+impl TryFrom<Amount> for SignedAmount {
+    type Error = ParseAmountError;
```
Comment (Kixunil) [inline units/src/amount.rs:1347]: This needs to be `OutOfRangeError`

File: units/src/amount.rs [hunk 4]
```diff
@@ -1835,6 +1852,48 @@ mod tests {
...
+        let sa_negative = SignedAmount(-123);
+        let ua_negative = Amount::try_from(sa_negative);
+        assert_eq!(
+            ua_negative,
+            Err(ParseAmountError::OutOfRange(OutOfRangeError {
+                is_signed: false,
+                is_greater_than_max: false
+            }))
+        );
+    }
```
Comment (Kixunil) [inline units/src/amount.rs:1895]: I don't think tests are important for such trivial code but I don't object to having them.

**Review by Kixunil — CHANGES_REQUESTED** [review]

**Comment by 0xkrieger** [comment]
can you suggest any other approach, I thought of doing
```rust
    fn try_from(value: SignedAmount) -> Result<Self, Self::Error> {
        let unsigned_amount = Amount(value.to_sat() as u64);
        Ok(unsigned_amount)
    }
```
but this will be wrong due to wraping-around of u64

File: units/src/amount.rs [hunk 5]
```diff
@@ -1344,10 +1344,10 @@ impl FromStr for SignedAmount {
 }
 
 impl TryFrom<Amount> for SignedAmount {
-    type Error = ParseAmountError;
+    type Error = OutOfRangeError;
 
     fn try_from(value: Amount) -> Result<Self, Self::Error> {
-        value.to_signed()
+        value.to_signed().map_err(|_| OutOfRangeError::too_big(true))
```
Comment (Kixunil) [inline units/src/amount.rs:1350]: Oh, `to_signed` should return the same error type.

**Review by Kixunil — DISMISSED** [review]
ACK f8145000270404809f5b88580c17b434a682208e

However maybe there's a value in squashing this. (And possibly change the error type but strictly speaking, it's out of scope of this PR.)

**Comment by Kixunil** [comment]
> but this will be wrong due to wraping-around of u64

I don't see the code you wrote in your comment in the PR.

File: units/src/amount.rs (hunk 5 above)
Comment (0xkrieger) [reply units/src/amount.rs:1350]: It doesn't [`to_signed()`](https://github.com/rust-bitcoin/rust-bitcoin/blob/53461f71c9cd6ed49acff61396758f53f8fa3de8/units/src/amount.rs#L867C1-L874C6)

**Comment by 0xkrieger** [comment]
> > but this will be wrong due to wraping-around of u64
> 
> I don't see the code you wrote in your comment in the PR.

I didn't implimented this approach 
```rust
    fn try_from(value: SignedAmount) -> Result<Self, Self::Error> {
        let unsigned_amount = Amount(value.to_sat() as u64);
        Ok(unsigned_amount)
    }
```
because it is failing the test

**Comment by 0xkrieger** [comment]
I will squash everything to one commit once it is done.

File: units/src/amount.rs [hunk 6]
```diff
@@ -1835,6 +1852,48 @@ mod tests {
...
+        );
+    }
+
+    #[test]
+    fn test_amount_try_from_signed_amount() {
+        let sa_positive = SignedAmount(123);
+        let ua_positive = Amount::try_from(sa_positive).unwrap();
+        assert_eq!(ua_positive, Amount::from_sat(123));
+
+        let sa_zero = SignedAmount(0);
```
Comment (yancyribbens) [inline units/src/amount.rs:1882]: same comment as above.

File: units/src/amount.rs [hunk 7]
```diff
@@ -1835,6 +1852,48 @@ mod tests {
...
+        let ua_zero = Amount::try_from(sa_zero).unwrap();
+        assert_eq!(ua_zero, Amount::ZERO);
+
+        let sa_negative = SignedAmount(-123);
+        let ua_negative = Amount::try_from(sa_negative);
```
Comment (yancyribbens) [inline units/src/amount.rs:1887]: can there be such a thing as `ua_negative`?  Amount can only be positive..

File: units/src/amount.rs [hunk 8]
```diff
@@ -1835,6 +1852,48 @@ mod tests {
...
+        let ua_positive = Amount::from_sat(123);
+        let sa_positive = SignedAmount::try_from(ua_positive).unwrap();
+        assert_eq!(sa_positive, SignedAmount(123));
+
+        let ua_zero = Amount::ZERO;
+        let sa_zero = SignedAmount::try_from(ua_zero).unwrap();
+        assert_eq!(sa_zero, SignedAmount(0));
+
+        let ua_max = Amount::MAX;
+        let sa_max = SignedAmount::try_from(ua_max);
```
Comment (yancyribbens) [inline units/src/amount.rs:1866]: should probably just be `error` or `e` or something, right?  This isn't really SignedAmount type anymore.

Comment (0xkrieger) [reply units/src/amount.rs:1866]: ok I will make all the test cases more clean and squash the commits.
Thanks for the review

File: units/src/amount.rs (hunk 5 above)
Comment (Kixunil) [reply units/src/amount.rs:1350]: I know, we should change it. Could be in this PR (ideally as the first commit) or in another one.

Comment (0xkrieger) [reply units/src/amount.rs]: I have squashed all the commits into one

File: units/src/amount.rs [hunk 9]
```diff
@@ -865,9 +865,9 @@ impl Amount {
     pub fn checked_rem(self, rhs: u64) -> Option<Amount> { self.0.checked_rem(rhs).map(Amount) }
 
     /// Convert to a signed amount.
-    pub fn to_signed(self) -> Result<SignedAmount, ParseAmountError> {
+    pub fn to_signed(self) -> Result<SignedAmount, OutOfRangeError> {
```
Comment (Kixunil) [inline units/src/amount.rs:868]: I meant to make this change a separate commit but don't worry about it, it's not important.

File: units/src/amount.rs [hunk 10]
```diff
@@ -957,6 +957,15 @@ impl FromStr for Amount {
     fn from_str(s: &str) -> Result<Self, Self::Err> { Amount::from_str_with_denomination(s) }
 }
 
+impl TryFrom<SignedAmount> for Amount {
+    type Error = OutOfRangeError;
+
+    fn try_from(value: SignedAmount) -> Result<Self, Self::Error> {
+        let unsigned_amount = value.to_unsigned()?;
```
Comment (Kixunil) [inline units/src/amount.rs:964]: Just `value.to_unsigned` should be fine. We had a clippy lint for this but it's disabled because of false positives. This one would've been true positive.

Comment (0xkrieger) [reply units/src/amount.rs]: Done

File: units/src/amount.rs [hunk 11]
```diff
@@ -1835,6 +1851,40 @@ mod tests {
...
+        let result = Amount::try_from(sa_negative);
+        assert_eq!(
+            result,
+            Err(OutOfRangeError {
+                is_signed: false,
+                is_greater_than_max: false
+            })
+        );
+    }
+
```
Comment (apoelstra) [inline units/src/amount.rs:1887]: In 668b11f8ae815e521be13ba243eaa75aaa828c90:

Trailing whitespace.

**Comment by apoelstra** [comment]
668b11f8ae815e521be13ba243eaa75aaa828c90 lgtm other than the trailing whitespace thing.

I think we should have a `TryFrom` in the other direction as well (or maybe just a `From`).

File: units/src/amount.rs (hunk 11 above)
Comment (0xkrieger) [reply units/src/amount.rs]: Removed it.

**Review by Kixunil — DISMISSED** [review]
ACK 3aefcb57d36f5e5d5d6e27c4ea92cd36d73a0326

**Review by tcharding — DISMISSED** [review]
ACK 3aefcb57d36f5e5d5d6e27c4ea92cd36d73a0326

**Comment by tcharding** [comment]
Ci fail is real
```
    --> units/src/amount.rs:1720:13
     |
1720 | /             if n1 <= i64::MAX as u64 {
1721 | |                 Ok(SignedAmount::from_sat(n1.try_into().unwrap()))
1722 | |             } else {
1723 | |                 Err(ParseAmountError::OutOfRange(OutOfRangeError::too_big(true)))
1724 | |             },
     | |_____________^ expected `amount::OutOfRangeError`, found `amount::ParseAmountError`
```

**Comment by 0xkrieger** [comment]
> Ci fail is real
> 
> ```
>     --> units/src/amount.rs:1720:13
>      |
> 1720 | /             if n1 <= i64::MAX as u64 {
> 1721 | |                 Ok(SignedAmount::from_sat(n1.try_into().unwrap()))
> 1722 | |             } else {
> 1723 | |                 Err(ParseAmountError::OutOfRange(OutOfRangeError::too_big(true)))
> 1724 | |             },
>      | |_____________^ expected `amount::OutOfRangeError`, found `amount::ParseAmountError`
> ```

Will resolve and push again shortly.

**Comment by Kixunil** [comment]
> I think we should have a `TryFrom` in the other direction as well (or maybe just a `From`).

What? There already are both impls (and `From` is impossible).

**Comment by tcharding** [comment]
Ha ha, I think Kix is right, I missed that while reviewing this morning :)

**Review by tcharding — APPROVED** [review]
ACK 251579f4eff65a7bd20e95d552e9cb6370e3043b

**Comment by apoelstra** [comment]
> > I think we should have a `TryFrom` in the other direction as well (or maybe just a `From`).
> 
> What? There already are both impls (and `From` is impossible).

Derp.

**Review by apoelstra — APPROVED** [review]
ACK 251579f4eff65a7bd20e95d552e9cb6370e3043b
